### ISSUE TITLE
RFC: A new kind of async

### ISSUE THREAD

**Issue by agnat** [nodejs/nan#715]
I always felt that the `AsyncWorker`s are suffering from a design defect. Right at its very core is a very subtle anti-pattern. I call it "composition by inheritance". Does that sound oxymoronic? Good. It always starts like this:

````
class AbstractSomething {
  virtual void doThing() = 0;
};
````

If you think about *adding a feature* by inheritance and about exposing both `AbstractSomething` and `FeatureSomething` to the user of your library, stop right there. You are almost certainly about to commit "composition by inheritance".

Another symptom of the anti-pattern is daisy-chaining. These class hierarchies never branch. Nobody wants to cut off a feature. Instead both sides, the library developers and the users, start accepting overhead. It becomes acceptable to tap into the class daisy chain at a point that gives you the features you need. If you inherit a couple of things you don't need, so be it.

It becomes acceptable because the alternative is a proliferation of `Something`s covering more feature permutations. This is so undesirable that we accept the overhead. Note, how the combinatory nature of our features has "toolbox" written all over it. Instead, "composition by inheritance" always creates a monolith. While the problem at hand requires loose coupling, our design uses the tightest coupling possible: inheritance.

So, what now? A few weeks back i started hacking on a little sketch exploring new ways to write asynchronous node addons. Beside loose coupling and a more toolboxy approach i was looking for a design that is familiar to node developers. In javascript it would seem natural to pass work functions and done handlers. So let's do this in C++.

C++ already has a concept named [callable](http://en.cppreference.com/w/cpp/concept/Callable). Relying on a language concept provides just the kind of loose coupling we are seeking. No inheritance, not even a real API. If it is callable, we're good.

Thinking about designs that take an arbitrary callable and execute it on the thread pool, another API comes to mind: Apple's grand central dispatch. Both environments have the notion of a main thread and they both have a thread pool. This thought gave the sketch its current direction and a name: [ncd – not central dispatch](https://github.com/agnat/ncd). Please take a look and let me know what you think. 

### From the (current) README:

````c++
void
eventEmittingWorker(Nan::FunctionCallbackInfo<Value> const& args) {
  using namespace std::string_literals;
  unsigned delay      = args[0]->Uint32Value();
  unsigned iterations = args[1]->Uint32Value();

  ncd::AsyncEventEmitter emitter(args[2].As<v8::Object>());  // ①

  ncd::defaultWorkQueue().dispatch([=](){                    // ②
    for (unsigned i = 0; i < iterations; ++i) {
      emitter.emit("progress"s, i);                          // ③
      usleep(delay);
    }
  }, std::bind(emitter.emit, "done"s));                      // ④
}
````

After grabbing some arguments the code creates an `AsyncEventEmitter` in (1). It wraps a javascript `EventEmitter` for use on a different thread. In (2) a lambda expression is dipatched to the thread pool. The expression captures copies of `delay`, `iterations` and the `emitter`. This is an ncd pattern: An `AsyncSomething` is first allocated on the main thread and then copied around to different threads. The call to `dispatch(...)` returns immediately and the lambda is launched on the thread pool. In (3) the async event emitter is invoked, sending progress events back to javascript. After the execution finishes the callback passed in (4) is invoked on the main thread. Here we just emit a done event.

The javascript code looks like this:

````javascript
const workers = require('workers')
  , EventEmitter = require('events')
  , ee = new EventEmitter()
    . on('progress', (p) => { console.log('progress', p) })
    . on('done',     ()  => { console.log('done') })

workers.eventEmittingWorker(10000, 10, ee)
````

**Comment by kkoopa** [comment]
I like the general idea. NAN itself is more of a toolbox than a monolithic framework. I have at some point referred to it as pick-and-mix. The `AsyncWorker` classes have a long legacy (and C++03 limitations) and are intended as a simple way of doing the most common tasks as well as a showcase of how to (correctly) combine libuv and node in user code.

I will start the discussion with a question: Why not use the std-like way of having free functions with an argument for an `Executor`, instead of the other way around? That is, why 
`Executor::dispatch` instead of `dispatch(..., Executor = defaultWorkQueue)`?

**Comment by agnat** [comment]
Hehe. That's an excellent question! Let me answer them both.

The implicit one first:

#### Why isn't the default queue the default queue?

One of the guiding principles is to use node conventions and patterns. In node the final done callback is always the last argument. The argument order of `dispatch(...)` follows this convention. If you look at the non-default case you'll probably agree that a trailing queue parameter is odd ... if not outright ugly.

#### Free vs. member function:

Indeed the choice here is arbitrary. In fact I looked at both variants and liked the free functions, too. I decided to keep the member functions because I _think_ it feels slightly more natural to javascript developers.

**Comment by agnat** [comment]
> The `AsyncWorker` classes [...] are intended as a simple way of doing the most common tasks as well as a showcase of how to (correctly) combine libuv and node in user code.

I agree... However, if I'm perfectly honest, whenever I invoke the statement, it feels a bit like an excuse. The first part acknowledges that the design doesn't scale, the second part encourages people to roll their own. The second part however does not work. Everything out there _is a_ `AsyncWorker`.

I believe this is holding us back. I believe removing the tight coupling will enable us to develop a new generation of tools. Take a look at the (experimental) [stream example](https://github.com/agnat/ncd/blob/master/examples/06.streams/streams.cpp). It features a full fledged node writable stream with back-pressure and all, that has a readable endpoint on a different thread. It's not finished but it looks very promising. Since node writable streams are event emitters this _already_ is very powerful. Now think about thread transformers...

**Comment by kkoopa** [comment]
One could have the `Executor / Queue` as the first argument, yet still have it optional through an overload set.
```c++
/* satisfies concept Executor */
struct DefaultWorkQueue = {
  void execute() { ... }
};
template <typename Executor, typename F, typename H>
dispatch(ExecutionPolicy policy, F callback, H emit);

template <typename F, typename H>
dispatch(F callback, H emit) { dispatch(DefaultWorkQueue(), std::forward(callback), std::forward(emit)); }
```

**Comment by agnat** [comment]
lol. No. I'm not against free functions. On the contrary. But I'd rather not add overloads, creating oddball signatures just to fold an argument. Explicit is good and NAN has a certain history of folding the wrong things... ;)

I also believe it makes sense to talk about this on a higher level first. Like, maybe I'm wrong. Maybe the reason why we don't see more asynchronous primitives has nothing to do with the design of `AsyncWorker`. Maybe it's because `AsyncWorker` actually covers 99% of the use cases and ncd is just a huge pile of wannabe C++ mumbo-jumbo that nobody needs.

I'd also love to hear from @mkrufky on this. You've been in there just recently and maybe could speak about how it is to maintain them workers?

Another question is how to release it. We don't _have_ to merge it into NAN. It also works as a companion library. One argument against a merger could be that ncd is very different from NAN proper. Not just style-wise, that too, but also on a deeper level.

If you look at line (3) of the OP, there clearly is some automatic conversion going on. Besides that the current implementation will not cut it and more work is needed, this is a change in paradigm. Maybe it's to confusing if parts of NAN offer automatic conversions and others don't. It also is quite a commitment...

These are just some ideas... Curious about your thoughts, questions and concerns regarding the bigger picture...

**Comment by kkoopa** [comment]
Since it uses modern features and such, I presume it could not be merged into NAN 2 as-is, but would have to wait for a future major release (if that happens). I always thought there would be more intermediate libraries built on top of NAN, since it primarily papers over version differences, but that never became too popular.

I speculate that the majority use case is: Copy boilerplate from some existing project, add own stuff, see that it runs and move on with other matters.

**Comment by mkrufky** [comment]
Also we should keep in mind that _ncd_ looks like it will require a minimum of `c++17`, while _nan_ itself is compatible with `c++03`.  When we drop support in _nan_ for node 0.10 & 0.12, then it only pushes up the requirement to `c++11`.  Better to keep _nan_ as backwards compatible as possible for as long as it makes sense.   Then again, I don't think it hurts to add *new* functionality to _nan_ that requires a newer compiler.  It just means that these new features wont necessarily be available without all of the proper requirements.

Back to _ncd_ - It really does look like something that might feel more natural to developers that want to create native bindings in newer. more flexible and interesting ways.  I think it's a great idea and you should definitely move forward with it.

**Comment by kkoopa** [comment]
Regarding C++ versions available: The worst compiler which can successfully compile any given version of Node (including V8) sets the bar. C++11 (g++ 4.8, VS 2013, clang whatever) for Node 4 through 8. Node 9 (because of V8) requires at least VS 2015 (V8 started using constexpr constructs that VS2013 cannot do). All features of NAN should work on all supported versions of Node.

**Comment by agnat** [comment]
Regarding the C++ standard: It's designed to only require C++11. If it does use anything else that's either an accident or me being lazy or childish. It evolved from a sketch...

Even a C++03 port seems feasible. Not that I would suggest that, but it is possible quite easily...

I will comment on the other subjects this weekend (I hope)...
